File: include/posix-extras-dbus/libqdbuscontext.zig
const QtC = @import("qt6zig");
const qtc = @import("qt6c");
const qdbuserror_enums = @import("libqdbuserror.zig").enums;

/// https://doc.qt.io/qt-6/qdbuscontext.html
pub const qdbuscontext = struct {
    /// New constructs a new QDBusContext object.
    ///
    ///
    pub fn New() QtC.QDBusContext {
        return qtc.QDBusContext_new();
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qdbuscontext.html#calledFromDBus)
    ///
    /// ``` self: QtC.QDBusContext ```
    pub fn CalledFromDBus(self: ?*anyopaque) bool {
        return qtc.QDBusContext_CalledFromDBus(@ptrCast(self));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qdbuscontext.html#connection)
    ///
    /// ``` self: QtC.QDBusContext ```
    pub fn Connection(self: ?*anyopaque) QtC.QDBusConnection {
        return qtc.QDBusContext_Connection(@ptrCast(self));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qdbuscontext.html#message)
    ///
    /// ``` self: QtC.QDBusContext ```
    pub fn Message(self: ?*anyopaque) QtC.QDBusMessage {
        return qtc.QDBusContext_Message(@ptrCast(self));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qdbuscontext.html#isDelayedReply)
    ///
    /// ``` self: QtC.QDBusContext ```
    pub fn IsDelayedReply(self: ?*anyopaque) bool {
        return qtc.QDBusContext_IsDelayedReply(@ptrCast(self));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qdbuscontext.html#setDelayedReply)
    ///
    /// ``` self: QtC.QDBusContext, enable: bool ```
    pub fn SetDelayedReply(self: ?*anyopaque, enable: bool) void {
        qtc.QDBusContext_SetDelayedReply(@ptrCast(self), enable);
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qdbuscontext.html#sendErrorReply)
    ///
    /// ``` self: QtC.QDBusContext, name: []const u8 ```
    pub fn SendErrorReply(self: ?*anyopaque, name: []const u8) void {
        const name_str = qtc.libqt_string{
            .len = name.len,
            .data = name.ptr,
        };
        qtc.QDBusContext_SendErrorReply(@ptrCast(self), name_str);
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qdbuscontext.html#sendErrorReply)
    ///
    /// ``` self: QtC.QDBusContext, typeVal: qdbuserror_enums.ErrorType ```
    pub fn SendErrorReply2(self: ?*anyopaque, typeVal: i32) void {
        qtc.QDBusContext_SendErrorReply2(@ptrCast(self), @intCast(typeVal));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qdbuscontext.html#sendErrorReply)
    ///
    /// ``` self: QtC.QDBusContext, name: []const u8, msg: []const u8 ```
    pub fn SendErrorReply22(self: ?*anyopaque, name: []const u8, msg: []const u8) void {
        const name_str = qtc.libqt_string{
            .len = name.len,
            .data = name.ptr,
        };
        const msg_str = qtc.libqt_string{
            .len = msg.len,
            .data = msg.ptr,
        };
        qtc.QDBusContext_SendErrorReply22(@ptrCast(self), name_str, msg_str);
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qdbuscontext.html#sendErrorReply)
    ///
    /// ``` self: QtC.QDBusContext, typeVal: qdbuserror_enums.ErrorType, msg: []const u8 ```
    pub fn SendErrorReply23(self: ?*anyopaque, typeVal: i32, msg: []const u8) void {
        const msg_str = qtc.libqt_string{
            .len = msg.len,
            .data = msg.ptr,
        };
        qtc.QDBusContext_SendErrorReply23(@ptrCast(self), @intCast(typeVal), msg_str);
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qdbuscontext.html#dtor.QDBusContext)
    ///
    /// Delete this object from C++ memory.
    ///
    /// ``` self: QtC.QDBusContext ```
    pub fn QDelete(self: ?*anyopaque) void {
        qtc.QDBusContext_Delete(@ptrCast(self));
    }
};

File: src_original/posix-extras-dbus/libqdbuscontext.cpp
#include <QDBusConnection>
#include <QDBusContext>
#include <QDBusMessage>
#include <QString>
#include <QByteArray>
#include <cstring>
#include <qdbuscontext.h>
#include "libqdbuscontext.h"
#include "libqdbuscontext.hxx"

QDBusContext* QDBusContext_new() {
    return new QDBusContext();
}

bool QDBusContext_CalledFromDBus(const QDBusContext* self) {
    return self->calledFromDBus();
}

QDBusConnection* QDBusContext_Connection(const QDBusContext* self) {
    return new QDBusConnection(self->connection());
}

QDBusMessage* QDBusContext_Message(const QDBusContext* self) {
    const QDBusMessage& _ret = self->message();
    // Cast returned reference into pointer
    return const_cast<QDBusMessage*>(&_ret);
}

bool QDBusContext_IsDelayedReply(const QDBusContext* self) {
    return self->isDelayedReply();
}

void QDBusContext_SetDelayedReply(const QDBusContext* self, bool enable) {
    self->setDelayedReply(enable);
}

void QDBusContext_SendErrorReply(const QDBusContext* self, const libqt_string name) {
    QString name_QString = QString::fromUtf8(name.data, name.len);
    self->sendErrorReply(name_QString);
}

void QDBusContext_SendErrorReply2(const QDBusContext* self, int typeVal) {
    self->sendErrorReply(static_cast<QDBusError::ErrorType>(typeVal));
}

void QDBusContext_SendErrorReply22(const QDBusContext* self, const libqt_string name, const libqt_string msg) {
    QString name_QString = QString::fromUtf8(name.data, name.len);
    QString msg_QString = QString::fromUtf8(msg.data, msg.len);
    self->sendErrorReply(name_QString, msg_QString);
}

void QDBusContext_SendErrorReply23(const QDBusContext* self, int typeVal, const libqt_string msg) {
    QString msg_QString = QString::fromUtf8(msg.data, msg.len);
    self->sendErrorReply(static_cast<QDBusError::ErrorType>(typeVal), msg_QString);
}

void QDBusContext_Delete(QDBusContext* self) {
    delete self;
}
